cpp-ch/local-engine/Shuffle/PartitionWriter.cpp

```diff
@@ -216,6 +213,82 @@ PartitionWriter::PartitionWriter(CachedShuffleWriter * shuffle_writer_)
     {
         partition_block_buffer.emplace_back(ColumnsBuffer(options->split_size));
     }
+    partition_buffer.resize(options->partition_nums);
+}
+CelebornPartitionWriter::CelebornPartitionWriter(CachedShuffleWriter * shuffleWriter, std::unique_ptr<CelebornClient> celeborn_client_)
+    : PartitionWriter(shuffleWriter), celeborn_client(std::move(celeborn_client_))
+{
+
+}
+
+void CelebornPartitionWriter::evictPartitions(bool for_memory_spill)
+{
+    auto spill_to_celeborn = [this]() -> void {
+        size_t partition_id = 0;
+        Stopwatch serialization_time_watch;
+        serialization_time_watch.start();
+        for (auto & partition : partition_buffer)
+        {
+            if (partition.empty()) continue;
+            WriteBufferFromOwnString output;
+            auto codec = DB::CompressionCodecFactory::instance().get(boost::to_upper_copy(shuffle_writer->options.compress_method), {});
+            CompressedWriteBuffer compressed_output(output, codec, shuffle_writer->options.io_buffer_size);
+            NativeWriter writer(compressed_output, 0, shuffle_writer->output_header);
+            for (const auto & block : partition)
+            {
+                writer.write(block);
+            }
+            compressed_output.sync();
+            Stopwatch push_time_watch;
+            push_time_watch.start();
+            celeborn_client->pushPartitionData(partition_id, output.str().data(), output.str().size());
+            shuffle_writer->split_result.partition_length[partition_id] += output.str().size();
+            partition_id++;
+            shuffle_writer->split_result.total_compress_time += compressed_output.getCompressTime();
+            shuffle_writer->split_result.total_write_time += compressed_output.getWriteTime();
+            shuffle_writer->split_result.total_write_time += push_time_watch.elapsedNanoseconds();
+            shuffle_writer->split_result.total_disk_time += push_time_watch.elapsedNanoseconds();
+        }
+        shuffle_writer->split_result.total_serialize_time += serialization_time_watch.elapsedNanoseconds();
+    };
+    Stopwatch spill_time_watch;
+    spill_time_watch.start();
+    if (for_memory_spill)
+    {
+        // escape memory track from current thread status; add untracked memory limit for create thread object, avoid trigger memory spill again
+        IgnoreMemoryTracker ignore(2 * 1024 * 1024);
+        ThreadFromGlobalPool thread(spill_to_celeborn);
+        thread.join();
+    }
+    else
+    {
+        spill_to_celeborn();
+    }
+    shuffle_writer->split_result.total_spill_time += spill_time_watch.elapsedNanoseconds();
+
+    for (auto & partition : partition_buffer)
+    {
+        partition.clear();
```

Comment:
hello, when I set config `spark.speculation true`,  and use celeborn shffule, some task would lead to core problem in this partition clear, and the exception as below
```
 Poco::Exception::Exception(String const&, int) @ 0x0000000013063bf9 in /data9/hadoop/yarn/local/usercache/zouyunhe/filecache/34062/libch.so
         > 1. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000b6a6b57 in /data9/hadoop/yarn/local/usercache/zouyunhe/filecache/34062/libch.so
         > 2. DB::Exception::createRuntime(int, String&) @ 0x000000000699694c in /data9/hadoop/yarn/local/usercache/zouyunhe/filecache/34062/libch.so
         > 3. void local_engine::safeCallVoidMethod<long>(JNIEnv_*, _jobject*, _jmethodID*, long) @ 0x000000000ba13f8d in /data9/hadoop/yarn/local/usercache/zouyunhe/filecache/34062/libch.so
         > 4. local_engine::ReservationListenerWrapper::free(long) @ 0x000000000ba13c29 in /data9/hadoop/yarn/local/usercache/zouyunhe/filecache/34062/libch.so
         > 5. CurrentMemoryTracker::free(long) @ 0x000000000b67d273 in /data9/hadoop/yarn/local/usercache/zouyunhe/filecache/34062/libch.so
         > 6. Allocator<false, false>::free(void*, unsigned long) @ 0x000000000b679be4 in /data9/hadoop/yarn/local/usercache/zouyunhe/filecache/34062/libch.so
         > 7. DB::ColumnString::~ColumnString() @ 0x000000001045e654 in /data9/hadoop/yarn/local/usercache/zouyunhe/filecache/34062/libch.so
         > 8. DB::ColumnNullable::~ColumnNullable() @ 0x000000001043f134 in /data9/hadoop/yarn/local/usercache/zouyunhe/filecache/34062/libch.so
         > 9. local_engine::CelebornPartitionWriter::evictPartitions(bool) @ 0x000000000b9e4a77 in /data9/hadoop/yarn/local/usercache/zouyunhe/filecache/34062/libch.so
         > 10. local_engine::CelebornPartitionWriter::stop() @ 0x000000000b9e5530 in /data9/hadoop/yarn/local/usercache/zouyunhe/filecache/34062/libch.so
         > 11. local_engine::CachedShuffleWriter::stop() @ 0x000000000b9d6b37 in /data9/hadoop/yarn/local/usercache/zouyunhe/filecache/34062/libch.so
         > 12. Java_io_glutenproject_vectorized_CHShuffleSplitterJniWrapper_stop @ 0x00000000069833c0 in /data9/hadoop/yarn/local/usercache
```

and would it be ok to remove this code to fix this problem? @liuneng1994

Reply:
We should find the root cause.

Reply:
this is the way to release memory, can't remove it